### PR TITLE
Setup a single top-level CORS middleware

### PR DESCRIPTION
**Pull request by bajtos**
This is a cleaned version of @claylo's pull request #174 which enables CORS across the entire loopback application, using reasonable defaults.

Close https://github.com/strongloop/loopback/issues/575

/to @raymondfeng or @ritch please review

### CODE REVIEW

File: templates/components/api-server/component.js [hunk 1]
```diff
@@ -67,10 +68,7 @@ template.server = {
           extended: true,
           limit: '100kb'
         },
-        cors: {
-          origin: true,
-          credentials: true
-        },
+        cors: false,
```
Comment (raymondfeng) [inline templates/components/api-server/component.js]: Is it to disable model level `cors` as we now have a global one?

File: templates/components/api-server/template/server/middleware.json [hunk 2]
```diff
@@ -3,7 +3,17 @@
...
   "initial": {
-    "compression": {}
+    "compression": {},
+    "cors": {
+      "params": {
+        "origin": true,
+        "credentials": true,
+        "exposedHeaders": true,
+        "methods": "GET,POST,PUT,PATCH,DELETE,OPTIONS",
+        "allowedHeaders": "Content-Type,Authorization",
```
Comment (raymondfeng) [inline templates/components/api-server/template/server/middleware.json]: Is it too restrictive? What about leaving it as default:
```
If not specified, defaults to reflecting the headers specified in the request's Access-Control-Request-Headers header.
```

File: templates/components/api-server/template/server/middleware.json [hunk 3]
```diff
@@ -3,7 +3,17 @@
...
   },
   "initial": {
-    "compression": {}
+    "compression": {},
+    "cors": {
+      "params": {
+        "origin": true,
+        "credentials": true,
+        "exposedHeaders": true,
+        "methods": "GET,POST,PUT,PATCH,DELETE,OPTIONS",
```
Comment (raymondfeng) [inline templates/components/api-server/template/server/middleware.json]: What about `HEAD`?

File: templates/components/api-server/template/server/middleware.json [hunk 4]
```diff
@@ -3,7 +3,17 @@
...
+        "allowedHeaders": "Content-Type,Authorization",
+        "maxAge": 180
```
Comment (raymondfeng) [inline templates/components/api-server/template/server/middleware.json]: Maybe we should set `maxAge` to a higher number as we don't change `CORS` policies dynamically.

File: templates/components/api-server/component.js (hunk 1 above)
Comment (bajtos) [inline templates/components/api-server/component.js]: Yes, see https://github.com/strongloop/strong-remoting/pull/165

**Comment by bajtos** [comment]
@claylo Could you please take a look at @raymondfeng's comments above regarding the settings you are proposing and explain why they are better than sticking with module's defaults?

Otherwise I am inclining towards using defaults as much as possible and reducing the scaffolded config to the following:

```json
{
  "origin": true,
  "credentials": true
}
```

Thoughts?

**Comment by claylo** [comment]
What I chose as the default config in my PR was what seemed like reasonable defaults at the time. Now that I'm looking back a few months later, I can't think of good arguments for keeping it that way.

Fully implemented CORS support would modify the CORS options on a per-endpoint basis, so that pre-flight OPTIONS requests to a particular endpoint actually returns the HTTP methods the endpoint supports given the authentication credentials provided. :smile: 

Since that's a bigger project than this, I think @bajtos suggestion to keep the scaffolded config to a minimum is a good idea. It's great that it's supported server-wide with this change, and I think further config can be left as an exercise for each application's implementation.

Thanks, guys!

**Comment by raymondfeng** [comment]
@bajtos Probably we should add `maxAge` to avoid aggressive preflight checks. See http://stackoverflow.com/questions/23543719/cors-access-control-max-age-is-ignored.

**Comment by bajtos** [comment]
@raymondfeng what value of `maxAge` do you propose? Supposedly browsers cap this value at different values (Chrome - 10 mins, Safari - 5 mins, Firefox - 24 hours).

**Comment by raymondfeng** [comment]
Set it to the highest value, such as `24 hours`.

**Comment by bajtos** [comment]
@raymondfeng fixed, LGTY now?

**Comment by raymondfeng** [comment]
LGTM

**Pull request merged**